File: schemakeeper-avro/src/main/scala/schemakeeper/avro/compatibility/CompatibilityType.scala
package schemakeeper.avro.compatibility

import enumeratum.EnumEntry.Snakecase
import enumeratum._

import scala.collection.immutable

sealed abstract class CompatibilityType extends EnumEntry with Snakecase

object CompatibilityType extends Enum[CompatibilityType] {
  override def values: immutable.IndexedSeq[CompatibilityType] = findValues

  case object None extends CompatibilityType

  case object Backward extends CompatibilityType

  case object Forward extends CompatibilityType

  case object Full extends CompatibilityType

  case object BackwardTransitive extends CompatibilityType

  case object ForwardTransitive extends CompatibilityType

  case object FullTransitive extends CompatibilityType
}

File: modules/common/src/main/java/schemakeeper/schema/CompatibilityType.java
package schemakeeper.schema;

import java.util.HashMap;
import java.util.Map;

public enum CompatibilityType {
    NONE("none"),
    BACKWARD("backward"),
    FORWARD("forward"),
    FULL("full"),
    BACKWARD_TRANSITIVE("backward_transitive"),
    FORWARD_TRANSITIVE("forward_transitive"),
    FULL_TRANSITIVE("full_transitive");

    private static final Map<String, CompatibilityType> nameToCompatibilityTypeMap;

    static {
        nameToCompatibilityTypeMap = new HashMap<>();
        nameToCompatibilityTypeMap.put("none", NONE);
        nameToCompatibilityTypeMap.put("backward", BACKWARD);
        nameToCompatibilityTypeMap.put("forward", FORWARD);
        nameToCompatibilityTypeMap.put("full", FULL);
        nameToCompatibilityTypeMap.put("backward_transitive", BACKWARD_TRANSITIVE);
        nameToCompatibilityTypeMap.put("forward_transitive", FORWARD_TRANSITIVE);
        nameToCompatibilityTypeMap.put("full_transitive", FULL_TRANSITIVE);
    }

    public final String identifier;

    CompatibilityType(String identifier) {
        this.identifier = identifier;
    }

    public static CompatibilityType findByName(String name) {
        return nameToCompatibilityTypeMap.getOrDefault(name.toLowerCase(), BACKWARD);
    }
}
